### DISCUSSION TITLE
Make ui.input's autocomplete usable on mobile phone

### DISCUSSION THREAD

**Discussion by rodja** [zauberzeug/nicegui#852] [Ideas / Feature Requests]
Currently the autocomplete feature of ui.input is unusable on mobile phone. What could be a good technique if there is no TAB button pressable?

**Comment by phoskee** [comment]
This doubt arose in your code review, and @falkoschindler too, and while I was fixing the code.
I had some time to think about it and thought that perhaps the solution might be to replace the _Tab keypress_ with the _Enter keypress_.

The question is, wouldn't this affect any `on('keypress.enter', foo)` entered by the user?

**Reply by falkoschindler** [reply]
"keypress.enter" could be an option. But it's sometimes used for submitting input fields. That might cause confusion.

**Reply by rodja** [reply]
Or `on('click', foo)`? But how to educate the users? In both cases it's not obvious what to do...

**Reply by phoskee** [reply]
@falkoschindler this is what I am referring to

**Comment by rodja** [comment]
Or we can maybe leverage the built-in HTML5 datalist element and the autocomplete attribute somehow?

**Reply by rodja** [reply]
I like the `<datalist>` approach most:

```py
    ui.html('''
        <datalist id="example-list">
            <option value="One"></option>
            <option value="Two"></option>
            <option value="three"></option>
        </datalist>
    ''')
    ui.input('Test').props('list="example-list"')
```

![Screen Shot 2023-04-28 at 11 12 43](https://user-images.githubusercontent.com/131391/235107495-ff825e48-3c32-49ba-b988-afed4d802625.png)

Especially if we combine it with the existing implementation I can imagine it to work very nicely.

**Reply by falkoschindler** [reply]
Very interesting! On desktop however, it is more a filtered dropdown list (like `ui.select` with input) than an autocompletion.

**Reply by rodja** [reply]
> On desktop however, it is more a filtered dropdown list (like ui.select with input) than an autocompletion.

Thats why I think we should combine the two. Either via media-query or simply allow both to coexist. Should work without problems. @phoskee would you like to try this?

**Reply by phoskee** [reply]
@rodja I am not very familiar with it but I will try! very nice idea!
Q: Assuming that `ui.html` will be populated by as many `<option value="x"></option>` as there are suggestions, what happens with a list of 1000 names with N people starting with A? Will we end up with N `<option value="One"></option>`?

**Reply by rodja** [reply]
@phoskee, yes, there must be an `<option>` for every suggestion. But as I understand it, the `<datalist>` can also be manipulated on the fly. So it may also possible to use a function which always generates the ten most likely matches and writes them as options to the datalist.

**Reply by phoskee** [reply]
@rodja Ok, I am referring to html, it is populated by 1000 `<option value="x"></option>`
as for the suggestions, I noticed that they are filtered with each input

**Reply by phoskee** [reply]
@rodja @falkoschindler  Tried with a list of many cities, it works!
Q;
- Should it be limited to mobile only? On Desktop it becomes a kind of `ui.select`
- Can be implemented optionally? like `input(autocomplete=x, mobile=True)`

**Reply by rodja** [reply]
> Should it be limited to mobile only? On Desktop it becomes a kind of ui.select

I think we should enable it by default for desktop. That will help with accessibility.

> Can be implemented optionally? like input(autocomplete=x, mobile=True)

I would vote for a parameter `popover` which is by default True but can be disabled to hide the popover on desktop. Setting popover to False should still use the datalist options on mobile but disable them for desktop.

**Reply by phoskee** [reply]
Not the best, for my opinion if I wanted something like that I would use `ui.select`

![Animation](https://user-images.githubusercontent.com/125825348/235134255-e4ddfcc6-be75-49b7-80f2-6b2e5c0b2f35.gif)

How do I know the type of device?
I am trying with some js functions but nothing

**Reply by rodja** [reply]
ChatGPT suggests

```js
  function isMobileDevice() {
    return (typeof window.orientation !== "undefined") || (navigator.userAgent.indexOf('IEMobile') !== -1);
  }
```

Or use a media-query to hide it for larger devices:

```css
/* Hide the default datalist appearance on devices with a minimum width of 768px */
@media (min-width: 768px) {
  input[list]::-webkit-calendar-picker-indicator {
    display: none;
  }
}
```

**Reply by phoskee** [reply]
Unfortunately as an Italian I do not have easy access to ChatGPT, however, I have found several solutions.
The least bad is this one:
```python
async def get_device():
                result = await ui.run_javascript('''
                    const detectDeviceType = () =>
                        /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(
                            navigator.userAgent
                        )
                            ? 'Mobile'
                            : 'Desktop';

                        detectDeviceType();
                ''', timeout=1)
                print(result)

                if result == 'Mobile':

                    options = ""
                    for opt in autocomplete:
                        options = options + ("<option value='{0}'></option>".format(opt))
                    ui.html('''
                        <datalist id="example-list">
                            {0}
                        </datalist>'''.format(options))
                else:
                    ui.html(' ')
```


media-query seems not work:

```python
ui.add_body_html('''
            <style>
                
                @media (min-width: 768px) {
                     input[list]::-webkit-calendar-picker-indicator {
                          display: none;
                   }
                 }
            </style>
            ''')
```

**Comment by rodja** [comment]
ChatGPT also suggested a combination of QInput, QSelect and QVirtualScroll. Something like this:

```vue
<template>
  <div>
    <q-input
      v-model="userInput"
      @input="filterSuggestions"
      label="Search"
      filled
    ></q-input>
    <q-select
      v-if="filteredSuggestions.length > 0"
      filled
      options-dense
      use-input
      v-model="userInput"
      :options="filteredSuggestions"
      behavior="dialog"
      @filter="filterSuggestions"
    >
      <template v-slot:before>
        <q-virtual-scroll
          :items="filteredSuggestions"
          :item-height="40"
          :items-size="filteredSuggestions.length"
          class="col-auto"
        >
          <template v-slot="{ item, index }">
            <q-item clickable @click.stop="userInput = item">
              <q-item-section>{{ item }}</q-item-section>
            </q-item>
          </template>
        </q-virtual-scroll>
      </template>
    </q-select>
  </div>
</template>
```

**Reply by falkoschindler** [reply]
This looks like re-inventing the autocompletion once again. We already have the `ui.select` with input field.

**Comment by phoskee** [comment]
My intention was to implement something simple that would give a little extra.

What do you think of a slot with an icon that appears along with the shadow-text? maybe too cumbersome..
Call up the function with the spacebar? Wouldn't work with an input of a phrase..

**Comment by phoskee** [comment]
With _"keypress.enter"_ seems to work without any problems:

```python
from nicegui import ui

def work():
    ui.notify('Some other task done')

options = ['Awesome', 'NiceGUI', 'AbCdE']
ui.input(autocomplete=options).on('keydown.enter', work)

ui.run()
```


![Animation](https://user-images.githubusercontent.com/125825348/235101788-70ed6a7b-e427-4e84-87ab-7080949664ba.gif)

**Reply by rodja** [reply]
Nice. So we would simply allow ENTER and TAB?

**Reply by falkoschindler** [reply]
Yes, with the downside of https://github.com/zauberzeug/nicegui/discussions/852#discussioncomment-5751734.

**Reply by phoskee** [reply]
No wait, only ENTER!

**Comment by falkoschindler** [comment]
After considering @phoskee's PR #879, which runs a JavaScript function for every `ui.input` to check if the browser is desktop or mobile, we tend towards a pure frontend implementation: We would wrap `q-input` with a custom Vue component and implement the desktop-mobile check right in JavaScript. This avoids async calls while constructing elements that depend on a client connection. And it resolves the possible conflict for the shared auto-index client which might be connected to desktop _and_ mobile devices.

**Reply by falkoschindler** [reply]
By the way: The wrapper could also implement the auto-completion in JavaScript. This would greatly reduce the latency, which is currently caused by sending user input and auto-suggestion back and forth between JavaScript and Python.

**Comment by falkoschindler** [comment]
This feature has been implemented in PR #1063.
